orttraining/orttraining/training_api/module.cc

```diff
@@ -20,6 +24,116 @@ namespace {
+    return Status::OK();
+  }
+
+  auto& inference_graph = inference_model->MainGraph();
+
+  InlinedVector<const NodeArg*> inference_graph_output_node_args;
+  inference_graph_output_node_args.reserve(inference_graph_outputs.size());
+  for (const auto& output_name : inference_graph_outputs) {
+    const NodeArg* output_node_arg = inference_graph.GetNodeArg(std::string(output_name));
+    ORT_RETURN_IF_NOT(output_node_arg, "Expected graph output for inference graph " + std::string(output_name) +
+                                           " could not be found. Please regenerate the eval graph.");
+    inference_graph_output_node_args.push_back(output_node_arg);
+  }
+
+  // Set the inference graph outputs, and remove any unused nodes.
+  inference_graph.SetOutputs(inference_graph_output_node_args);
+  ORT_RETURN_IF_ERROR(RemoveUnusedNodes(inference_graph, inference_graph_output_node_args));
+
+  ORT_RETURN_IF_ERROR(inference_graph.Resolve());
+
+  return Status::OK();
+}
+
+InlinedVector<std::string_view> ParseInferenceGraphOutputsFromMetadata(const ONNX_NAMESPACE::ModelProto& model_proto) {
+  std::string inference_graph_outputs_string;
+  for (const auto& model_metadata : model_proto.metadata_props()) {
+    if (model_metadata.key() == Module::InferenceGraphOutputsMetadataString) {
+      inference_graph_outputs_string = model_metadata.value();
+      break;
+    }
+  }
+
+  return onnxruntime::utils::SplitString(inference_graph_outputs_string, ",");
+}
+
+Status TransformModelInputsForInference(std::shared_ptr<Model> inference_model, const std::unordered_map<std::string, std::shared_ptr<Parameter>>& named_parameters, const DataTransferManager& data_transfer_manager) {
+  std::vector<const NodeArg*> user_graph_inputs;
+  for (auto& graph_input_node_arg : inference_model->MainGraph().GetInputs()) {
+    auto named_parameter_it = named_parameters.find(graph_input_node_arg->Name());
+    if (named_parameter_it == named_parameters.end()) {
+      if (inference_model->MainGraph().GetConsumerNodes(graph_input_node_arg->Name()).empty()) {
+        continue;
+      }
+      user_graph_inputs.emplace_back(graph_input_node_arg);
+    } else {
+      const OrtMemoryInfo cpu_alloc_info{onnxruntime::CPU, OrtDeviceAllocator};
+      const Tensor& src_tensor = named_parameter_it->second->Data().Get<onnxruntime::Tensor>();
+      InlinedVector<char> tensor_data_buffer{};
+      tensor_data_buffer.resize(src_tensor.SizeInBytes());
+      gsl::span<char> dst_span = gsl::make_span(tensor_data_buffer);
+      ORT_RETURN_IF_NOT(src_tensor.SizeInBytes() == static_cast<size_t>(dst_span.size_bytes()), "src size != dst size");
+      Tensor dst_tensor{src_tensor.DataType(), src_tensor.Shape(), dst_span.data(), cpu_alloc_info};
+      ORT_RETURN_IF_ERROR(data_transfer_manager.CopyTensor(src_tensor, dst_tensor));
+      inference_model->MainGraph().AddInitializedTensor(onnxruntime::utils::TensorToTensorProto(dst_tensor, named_parameter_it->first));
```

Comment:
to be safety, maybe we should check if the intiializer already exists. Ideally it should NOT exit, am I right?

Reply:
The initializer should not exist. All model parameters should be graph inputs and not be a part of the initializers. I can add a check.